core/src/main/java/kafka/server/share/SharePartitionManager.java

```diff
@@ -564,16 +592,29 @@ void maybeProcessFetchQueue() {
                         );
                     } else {
                         sharePartition.releaseFetchLock();
-                        log.info("Record lock partition limit exceeded for SharePartition with key {}, " +
-                            "cannot acquire more records", sharePartitionKey);
                     }
                 }
             });
 
-            if (topicPartitionData.isEmpty()) {
-                // No locks for share partitions could be acquired, so we complete the request and
-                // will re-fetch for the client in next poll.
+            if (shareFetchPartitionData.partitionMaxBytes.isEmpty()) {
+                // If there are no partitions to fetch then complete the future with an empty map.
                 shareFetchPartitionData.future.complete(Collections.emptyMap());
+                // Release the lock so that other threads can process the queue.
+                releaseProcessFetchQueueLock();
+                if (!fetchQueue.isEmpty())
+                    maybeProcessFetchQueue();
+                return;
+            }
+            if (topicPartitionData.isEmpty()) {
+                // No locks for any of the share partitions in the fetch request could be acquired.
+                Set<Object> delayedShareFetchWatchKeys = new HashSet<>();
+                shareFetchPartitionData.partitionMaxBytes.keySet().forEach(
+                    topicIdPartition -> delayedShareFetchWatchKeys.add(
+                        new DelayedShareFetchKey(topicIdPartition, shareFetchPartitionData.groupId)));
+
+                // Add the share fetch to the delayed share fetch purgatory to process the fetch request.
+                addDelayedShareFetch(new DelayedShareFetch(shareFetchPartitionData, replicaManager, partitionCacheMap),
```

Comment:
`delayedShareFetchPurgatory.tryCompleteElseWatch()` calls `tryComplete`, which tries to acquire the partition level lock. Could we remove the logic for acquiring the partition lock in this method?

Reply:
IIUC, we do not want that on adding an entry to the purgatory, it should try to acquire the partition level lock to check if we can acquire records in the partition because we have already tried to acquire the lock while processing the fetch queue and failed .
There is no method in `DelayedOperationPurgatory` by which we can do an operation like add an operation and keys associated to it and just watch it without trying. So, I have introduced a variable `isTryingForFirstTime` in `DelayedShareFetch` where once you add an entry to the purgatory the corresponding tryComplete won't acquire the partition lock.

Reply:
I mean that `DelayedShareFetch.tryComplete()` needs to implement the logic for acquiring the partition level lock for both the initial `delayedShareFetchPurgatory.tryCompleteElseWatch()` call and the subsequent `delayedShareFetchPurgatory.checkAndComplete()` calls. Instead of implementing the lock acquiring logic in both `SharePartitionManager` and `DelayedShareFetch`, could we just do it once in `DelayedShareFetch`? Then, we don't need the `isTryingForFirstTime` flag.

> There is no method in DelayedOperationPurgatory by which we can do an operation like add an operation and keys associated to it and just watch it without trying. 

That's true. My suggestion is to not acquire the lock in `SharePartitionManager` and do that in `delayedShareFetchPurgatory.tryCompleteElseWatch()` (through `DelayedShareFetch.tryComplete()`) instead.

Reply:
> My suggestion is to not acquire the lock in SharePartitionManager and do that in delayedShareFetchPurgatory.tryCompleteElseWatch() (through DelayedShareFetch.tryComplete()) instead.

Hi @junrao , I don't think we can do that, because in `SharePartitionManager`, we do a `replicaManager.fetchMessages` to avoid a tight loop between the broker and share consumer. However, we do a `replicaManager.readFromLog` in the onComplete() of `DelayedShareFetch`. So, if we do a tryComplete and the partitions get acquired, then we would be executing the `replicaManager.readFromLog` right after `tryComplete` which will cause a tight loop between broker and share consumer.

Reply:
@adixitconfluent : I am not sure about the usage of `replicaManager.fetchMessages`. The intention seems to use it for satisfying minBytes. However, the logic to unblock a pending fetch request is not the same for share fetch (e.g. it doesn't take into consideration of the state of the fetched batches). So, I am not sure if we could reuse fetch purgatory for share fetch. Also, it's probably cleaner to use a single purgatory for pending share fetch requests instead of two.

> we want to ensure that ShareFetch requests do not immediately return if they hit the partition lock limit.

If this is the goal for this PR instead of satisfying minBytes, we could change the flow in SharePartitionManager to get rid of `replicaManager.fetchMessages` and always call `delayedShareFetchPurgatory.tryCompleteElseWatch()`. When the delayed operation can be completed, rely on `DelayedShareFetch.onComplete()` to call `replicaManager.readFromLog()` to obtain the data. We can figure out how to implement minBytes on share fetch purgatory later.

Reply:
Hi @junrao, I have changed the flow in SharePartitionManager to get rid of `replicaManager.fetchMessages` and always call `delayedShareFetchPurgatory.tryCompleteElseWatch()`. Sorry for the delayed response. Please review it when you can. Thanks!